
<a id="overview"></a>

# Exploring Walmart Sales Data


## Overview
In this assignment, you will work with sample Walmart data looking at weekly sales in order to project sales for different departments in each store. One of the biggest challenges is the fact that we have limited data. Your task will be to import raw data, clean the data, and structure your insights by looking at historical sales data from nearly 50 Walmart stores around the United States (and in every store, there are multiple departments).


**Expected Time to complete: 4 hours**

## Objectives

This assignment will require you to:

- Read/write CSV files using Python's `pandas` module.
- Apply basic pandas functionality like `merge`, `cut`, filtering, etc.
- Practice graphing trends to ascertain insights from them.
- Clean and structure Walmart sales data in order to forecast demand.


## Problem

Your job is to project sales for each department in every store. Easy, right? You'll be answering questions, such as:

- What are the average weekly sales by region?
- Is there a correlation between fuel price and sales?

As you navigate the notebook, you will see clearly labeled sections setting up questions for you to solve marked **required.** You will need to attempt answers for all of these **required** questions. Please include all work within your Jupyter notebook.


## Instructions

1. Open the assignment notebook.
1. Save a copy of your notebook and retitle it: "yourname_assignment.ipynb"
1. Attempt answers for all required, graded questions. Include comments to explain your logic.
1. Submit your notebook to your instructional team by the due date.


## Data
Our data has been sampled from an [open-source project hosted by Walmart on Kaggle.com](https://www.kaggle.com/c/walmart-recruiting-store-sales-forecasting/overview). When you click into the data folder, you'll see several files:

Within `train.csv` we have:
- Store - the store number
- Dept - the department number
- Date - the week
- `Weekly_Sales` -  sales for the given department in the given store
- `IsHoliday` - whether the week is a special holiday week

in `stores.csv` we have data related to specific stores, departments, and regional activity for the given dates:

- Store - the store number
- Type - string
- Size - size of the store

in `features.csv` we have data related to specific stores, departments, and regional activity for the given dates:

- Store - the store number
- Date - the week
- Temperature - average temperature in the region
- `Fuel_Pric`e - cost of fuel in the region
- `MarkDown1-5` - anonymized data related to promotional markdowns that Walmart is running. MarkDown data is only - available after Nov 2011, and is not available for all stores all the time. Any missing value is marked with an NA.
- CPI - the consumer price index
- Unemployment - the unemployment rate
- `IsHoliday` - whether the week is a special holiday week

in `test.csv` we have a sample of weeks that we can attempt to forecast weekly sales for (and then test our prediction using Mean Squared Error).

---

### Setting up our Environment

In [119]:
# Import the requisite packages
# Pandas for data manipulation
import pandas as pd

# Matplotlib for basic plotting
import matplotlib

# Ignore potential warnings
import warnings
warnings.filterwarnings("ignore")


In [121]:
# Now let's read in the **Training** data and take a look at the first few rows
training_df = pd.read_csv("data/train.csv")

print(training_df.shape)

training_df.head()


(421570, 5)


Store  Dept        Date  Weekly_Sales  IsHoliday
0      1     1  2010-02-05      24924.50      False
1      1     1  2010-02-12      46039.49       True
2      1     1  2010-02-19      41595.55      False
3      1     1  2010-02-26      19403.54      False
4      1     1  2010-03-05      21827.90      False

In [123]:
# Next, read in the **Features** dataset
features = pd.read_csv("data/features.csv")
features.head()

Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0      1  2010-02-05        42.31       2.572        NaN        NaN   
1      1  2010-02-12        38.51       2.548        NaN        NaN   
2      1  2010-02-19        39.93       2.514        NaN        NaN   
3      1  2010-02-26        46.63       2.561        NaN        NaN   
4      1  2010-03-05        46.50       2.625        NaN        NaN   

   MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        NaN        NaN        NaN  211.096358         8.106      False  
1        NaN        NaN        NaN  211.242170         8.106       True  
2        NaN        NaN        NaN  211.289143         8.106      False  
3        NaN        NaN        NaN  211.319643         8.106      False  
4        NaN        NaN        NaN  211.350143         8.106      False

In [125]:
# Finally, read in the **Stores** dataset
stores = pd.read_csv("data/stores.csv")
stores.head()

Store Type    Size
0      1    A  151315
1      2    A  202307
2      3    B   37392
3      4    A  205863
4      5    B   34875

---

# REQUIRED / GRADED
> These questions are required for submission and will be graded.

### Tutorial: 

A Brief Primer on Merging 
> The following example was sourced from http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python.


- **Natural join**: To keep only rows that match from the data frames, specify the argument how=‘inner’
- **Full outer join**:To keep all rows from both data frames, specify how=‘outer’
- **Left outer join**:To include all the rows of your data frame x and only those from y that match, specify how=‘left’
- **Right outer join**:To include all the rows of your data frame y and only those from x that match, specify how=‘right’.

![alt text](http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png)

For another great resource on merging with pandas, [click here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html).


---

## Question 1

Merge the three DataFrames together into a single dataframe called `training_df`.

> Hint, the resulting dataset should be of shape `(421570, 17)`

In [127]:
# Merge dataframes into single dataframe

training_df = pd.merge(training_df, stores, on = ["Store"])
training_df = pd.merge(training_df, features, on = ["Store", "Date"])

# Merge training_df (which now has features in it) and stores

# Test csv merge
training_df.head()


Store  Dept        Date  Weekly_Sales  IsHoliday_x Type    Size  \
0      1     1  2010-02-05      24924.50        False    A  151315   
1      1     1  2010-02-12      46039.49         True    A  151315   
2      1     1  2010-02-19      41595.55        False    A  151315   
3      1     1  2010-02-26      19403.54        False    A  151315   
4      1     1  2010-03-05      21827.90        False    A  151315   

   Temperature  Fuel_Price  MarkDown1  MarkDown2  MarkDown3  MarkDown4  \
0        42.31       2.572        NaN        NaN        NaN        NaN   
1        38.51       2.548        NaN        NaN        NaN        NaN   
2        39.93       2.514        NaN        NaN        NaN        NaN   
3        46.63       2.561        NaN        NaN        NaN        NaN   
4        46.50       2.625        NaN        NaN        NaN        NaN   

   MarkDown5         CPI  Unemployment  IsHoliday_y  
0        NaN  211.096358         8.106        False  
1        NaN  211.242170         8.106         True  
2        NaN  211.289143         8.106        False  
3        NaN  211.319643         8.106        False  
4        NaN  211.350143         8.106        False

In [131]:
# Checking dataset shape

print(training_df.shape)

(421570, 17)


---

### Tutorial: 

Once we have our new dataframe, let's do a little exploring!

In [ ]:
training_df.head(2)

In [133]:
# Note that `IsHoliday_x` and `IsHoliday_y` are repeats from two different datasets; let's remove one of them and rename the other.

# We'll remove the `IsHoliday_x` coumn
del training_df['IsHoliday_x']

# Create a new column called `isHoliday`
training_df['isHoliday'] = training_df['IsHoliday_y']

# And delete the `IsHoliday_y` column
del training_df['IsHoliday_y']

training_df.head()

Store  Dept        Date  Weekly_Sales Type    Size  Temperature  \
0      1     1  2010-02-05      24924.50    A  151315        42.31   
1      1     1  2010-02-12      46039.49    A  151315        38.51   
2      1     1  2010-02-19      41595.55    A  151315        39.93   
3      1     1  2010-02-26      19403.54    A  151315        46.63   
4      1     1  2010-03-05      21827.90    A  151315        46.50   

   Fuel_Price  MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5  \
0       2.572        NaN        NaN        NaN        NaN        NaN   
1       2.548        NaN        NaN        NaN        NaN        NaN   
2       2.514        NaN        NaN        NaN        NaN        NaN   
3       2.561        NaN        NaN        NaN        NaN        NaN   
4       2.625        NaN        NaN        NaN        NaN        NaN   

          CPI  Unemployment  isHoliday  
0  211.096358         8.106      False  
1  211.242170         8.106       True  
2  211.289143         8.106      False  
3  211.319643         8.106      False  
4  211.350143         8.106      False

In [135]:
# Now let's run some basic descriptive statistics
training_df.describe()

Store           Dept   Weekly_Sales           Size  \
count  421570.000000  421570.000000  421570.000000  421570.000000   
mean       22.200546      44.260317   15981.258123  136727.915739   
std        12.785297      30.492054   22711.183519   60980.583328   
min         1.000000       1.000000   -4988.940000   34875.000000   
25%        11.000000      18.000000    2079.650000   93638.000000   
50%        22.000000      37.000000    7612.030000  140167.000000   
75%        33.000000      74.000000   20205.852500  202505.000000   
max        45.000000      99.000000  693099.360000  219622.000000   

         Temperature     Fuel_Price      MarkDown1      MarkDown2  \
count  421570.000000  421570.000000  150681.000000  111248.000000   
mean       60.090059       3.361027    7246.420196    3334.628621   
std        18.447931       0.458515    8291.221345    9475.357325   
min        -2.060000       2.472000       0.270000    -265.760000   
25%        46.680000       2.933000    2240.270000      41.600000   
50%        62.090000       3.452000    5347.450000     192.000000   
75%        74.280000       3.738000    9210.900000    1926.940000   
max       100.140000       4.468000   88646.760000  104519.540000   

           MarkDown3      MarkDown4      MarkDown5            CPI  \
count  137091.000000  134967.000000  151432.000000  421570.000000   
mean     1439.421384    3383.168256    4628.975079     171.201947   
std      9623.078290    6292.384031    5962.887455      39.159276   
min       -29.100000       0.220000     135.160000     126.064000   
25%         5.080000     504.220000    1878.440000     132.022667   
50%        24.600000    1481.310000    3359.450000     182.318780   
75%       103.990000    3595.040000    5563.800000     212.416993   
max    141630.610000   67474.850000  108519.280000     227.232807   

        Unemployment  
count  421570.000000  
mean        7.960289  
std         1.863296  
min         3.879000  
25%         6.891000  
50%         7.866000  
75%         8.572000  
max        14.313000

In [137]:
# Note: Why don't we see `isHoliday` in here? That's because the dataframe thinks that "True/False" column data are string values.
# We can test this by calling `describe` on the `isHoliday` column. When the results show "count, unique, etc" (instead of "mean, min, max" etc) that means Pandas is interpreting the data as a string.

training_df['isHoliday'].describe()

count     421570
unique         2
top        False
freq      391909
Name: isHoliday, dtype: object

---

### Tutorial

Great work! Before we go any further, let's look at a brief primer on cuts.

In the next portion of our assignment, you'll need to get familiar with using Pandas to cut and bin data. Here are a few examples:


```
pandas.cut(pandas.Series, bins)

where bins is usually an int or a list
```

If bins is of type `int`, then the `series` is replaced with intervals of roughly equal size that segment the data between the `min` of the Series and the `max` of the Series.

For example, if we did:

```
pd.cut(training_df['CPI'], 4).vlaue_counts()
```

then we'd get the following:

```
(125.963, 151.356]    207891
(201.941, 227.233]    154507
(176.648, 201.941]     59172
(151.356, 176.648]         0
```

Note that the lowest interval range starts at .1% less than the min (125.963) and the largest interval ends at .1% higher than the largest value in CPI (227.233):

```
training_df['CPI'].describe()
----
count    421570.000000
mean        171.201947
std          39.159276
min         126.064000
25%         132.022667
50%         182.318780
75%         212.416993
max         227.232807
```

> For more information, see [Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html).

Also note that we have a bin interval with `0` values in it. This is because the intervals are created solely to be of equal length between the `min` and the `max` of the series. If that means an interval will exist with no values in it, then so be it!

But if we use a `list` for bins, then we can dictate the intervals that we want to use:

```
pd.cut(training_df['CPI'], [100, 150, 200, 220, 250]).value_counts()
---
(100, 150]    207891
(200, 220]    115422
(150, 200]     59172
(220, 250]     39085
Name: CPI, dtype: int64
```

**Warning:** You can also define intervals that exclude rows:

```

pd.cut(training_df['CPI'], [50, 100, 200])
---
Date
2010-02-05               NaN
2010-02-12               NaN
2010-02-19               NaN
2010-02-26               NaN
2010-03-05               NaN
                   ...      
2012-09-28    (100.0, 200.0]
2012-10-05    (100.0, 200.0]
2012-10-12    (100.0, 200.0]
2012-10-19    (100.0, 200.0]
2012-10-26    (100.0, 200.0]
Name: CPI, Length: 421570, dtype: category
Categories (2, interval[int64]): [(50, 100] < (100, 200]]
```

So be careful!


Finally, as an optional parameter, we can define a `label` for each category to make these a bit more readable:

```
pd.cut(training_df['CPI'], 4, labels=['low CPI', 'low-medium CPI',  'medium CPI', 'high CPI']).value_counts()
---
low CPI           207891
high CPI          154507
medium CPI         59172
low-medium CPI         0
Name: CPI, dtype: int64
```

---

## Question 2

Use `pandas.cut` for each of the following questions:

- Part 1. Cut Weekly Sales into 4 bins of equal length (the bins should have the same size as one another)
- Part 2. Cut Weekly Sales into 4 bins of equal size (the bins should have roughly the same number of elements in them)
- Part 3. Create a new column called `Sales_Categorical` that has one of 4 values:
    - "Negative" if Weekly Sales <= 0
    - "Low" if 0 < Weekly Sales <= 10000
    - "Moderate" if 10000 < Weekly Sales <= 20000
    - "High" if 20000 < Weekly Sales


```
training_df['Sales_Categorical'].value_counts() should be:

Low         238587
High        106452
Moderate     75173
Negative      1358
Name: Sales_Categorical, dtype: int64
```
    

In [52]:
# Understand Weekly Sales data to build bins

training_df['Weekly_Sales'].describe()

count    421570.000000
mean      15981.258123
std       22711.183519
min       -4988.940000
25%        2079.650000
50%        7612.030000
75%       20205.852500
max      693099.360000
Name: Weekly_Sales, dtype: float64

In [58]:
# Part 1
# Bins of equal length

pd.cut(training_df['Weekly_Sales'], 4).value_counts()

Weekly_Sales
(-5687.028, 169533.135]    421050
(169533.135, 344055.21]       495
(344055.21, 518577.285]        21
(518577.285, 693099.36]         4
Name: count, dtype: int64

In [62]:
# Part 2
# Bins with equal number of elements

pd.cut(training_df['Weekly_Sales'], [-4988.94, 2079.65, 7612.03, 20205.8525, 693099.36]).value_counts()

Weekly_Sales
(2079.65, 7612.03]        105393
(20205.852, 693099.36]    105393
(-4988.94, 2079.65]       105392
(7612.03, 20205.852]      105391
Name: count, dtype: int64

In [178]:
# Part 3
# New column - Sales Categorical

# Establishing parameters for value categories
def Sales_Categorical(value):
    if value <= 0:
        return 'Negative'
    elif value <= 10000:
        return 'Low'
    elif value <= 20000:
        return 'Moderate'
    else:
        return 'High'

# Transposing to new column
training_df['Sales_Categorical'] = training_df['Weekly_Sales'].apply(Sales_Categorical)

# Check work
print(training_df[['Weekly_Sales', 'Sales_Categorical']].head(5))

   Weekly_Sales Sales_Categorical
0      24924.50              High
1      46039.49              High
2      41595.55              High
3      19403.54          Moderate
4      21827.90              High


---

### Tutorial: .loc vs .iloc

Onward! Next up, we'll look at one of the most commonly mixed up definitions in Pandas: `loc` vs. `iloc`:

- `loc` returns rows (or columns) with particular **labels** from the index.
- `iloc` returns rows (or columns) at particular **positions** in the index (which means `iloc` only takes integers as a parameter).

Here's a fun mnemonic:  
- .**Iloc** can only take **Integers** but 
- .**Loc** can use **Labels**

> Hint: For more examples, check out the official Pandas documentation for [.loc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) and [.iloc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html).

In [180]:
# Grab the first 3 elements (by looking at the index)
training_df.loc[:2]

Store  Dept        Date  Weekly_Sales Type    Size  Temperature  \
0      1     1  2010-02-05      24924.50    A  151315        42.31   
1      1     1  2010-02-12      46039.49    A  151315        38.51   
2      1     1  2010-02-19      41595.55    A  151315        39.93   

   Fuel_Price  MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5  \
0       2.572        NaN        NaN        NaN        NaN        NaN   
1       2.548        NaN        NaN        NaN        NaN        NaN   
2       2.514        NaN        NaN        NaN        NaN        NaN   

          CPI  Unemployment  isHoliday Sales_Categorical  
0  211.096358         8.106      False              High  
1  211.242170         8.106       True              High  
2  211.289143         8.106      False              High

In [182]:
# Grab the first 3 elements (no matter what the index is)
training_df.iloc[:2]

Store  Dept        Date  Weekly_Sales Type    Size  Temperature  \
0      1     1  2010-02-05      24924.50    A  151315        42.31   
1      1     1  2010-02-12      46039.49    A  151315        38.51   

   Fuel_Price  MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5  \
0       2.572        NaN        NaN        NaN        NaN        NaN   
1       2.548        NaN        NaN        NaN        NaN        NaN   

          CPI  Unemployment  isHoliday Sales_Categorical  
0  211.096358         8.106      False              High  
1  211.242170         8.106       True              High

In [184]:
# Let's change the index to see this in action!

# Change the index to be a pandas `Datetime` index:
training_df.index = pd.to_datetime(training_df['Date'])
training_df = training_df.sort_index()

# Remove the `date` column (we won't need it anymore):
del training_df['Date']

training_df.head(2)

Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-02-05      1     1      24924.50    A  151315        42.31       2.572   
2010-02-05     29     5      15552.08    B   93638        24.36       2.788   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-02-05        NaN        NaN        NaN        NaN        NaN  211.096358   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-02-05         8.106      False              High  
2010-02-05        10.064      False          Moderate

In [186]:
training_df.iloc[:2]  # this is still the same because we are using "positional" location

Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-02-05      1     1      24924.50    A  151315        42.31       2.572   
2010-02-05     29     5      15552.08    B   93638        24.36       2.788   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-02-05        NaN        NaN        NaN        NaN        NaN  211.096358   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-02-05         8.106      False              High  
2010-02-05        10.064      False          Moderate

In [188]:
try:
    training_df.loc[:2]  # Note that this throws an error now because the labels of the index are not integers!
except Exception as e:
    print(e)

cannot do slice indexing on DatetimeIndex with these indexers [2] of type int


In [190]:
# But now we can use the date index to filter quickly by dates:
training_df.loc["2010-02-05"]  # Get only values from February 5th, 2010

Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-02-05      1     1      24924.50    A  151315        42.31       2.572   
2010-02-05     29     5      15552.08    B   93638        24.36       2.788   
2010-02-05     29     6       3200.22    B   93638        24.36       2.788   
2010-02-05     29     7      10820.05    B   93638        24.36       2.788   
2010-02-05     29     8      20055.64    B   93638        24.36       2.788   
...           ...   ...           ...  ...     ...          ...         ...   
2010-02-05     43     9         16.42    C   41062        47.31       2.572   
2010-02-05     43    60        871.20    C   41062        47.31       2.572   
2010-02-05     45    31       3384.25    B  118221        27.31       2.784   
2010-02-05     43    59        410.84    C   41062        47.31       2.572   
2010-02-05     45    34      15134.00    B  118221        27.31       2.784   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-02-05        NaN        NaN        NaN        NaN        NaN  211.096358   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
...               ...        ...        ...        ...        ...         ...   
2010-02-05        NaN        NaN        NaN        NaN        NaN  203.064274   
2010-02-05        NaN        NaN        NaN        NaN        NaN  203.064274   
2010-02-05        NaN        NaN        NaN        NaN        NaN  181.871190   
2010-02-05        NaN        NaN        NaN        NaN        NaN  203.064274   
2010-02-05        NaN        NaN        NaN        NaN        NaN  181.871190   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-02-05         8.106      False              High  
2010-02-05        10.064      False          Moderate  
2010-02-05        10.064      False               Low  
2010-02-05        10.064      False          Moderate  
2010-02-05        10.064      False              High  
...                  ...        ...               ...  
2010-02-05         9.521      False               Low  
2010-02-05         9.521      False               Low  
2010-02-05         8.992      False               Low  
2010-02-05         9.521      False               Low  
2010-02-05         8.992      False          Moderate  

[2955 rows x 16 columns]

In [192]:
training_df.loc["2010-02-05":"2010-03-02"]  # Grab values on a range of dates

Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-02-05      1     1      24924.50    A  151315        42.31       2.572   
2010-02-05     29     5      15552.08    B   93638        24.36       2.788   
2010-02-05     29     6       3200.22    B   93638        24.36       2.788   
2010-02-05     29     7      10820.05    B   93638        24.36       2.788   
2010-02-05     29     8      20055.64    B   93638        24.36       2.788   
...           ...   ...           ...  ...     ...          ...         ...   
2010-02-26     41    91      49688.56    A  196321        21.84       2.586   
2010-02-26     12     5      28498.57    B  112238        50.23       2.825   
2010-02-26     17    81      10706.26    B   93188        15.64       2.667   
2010-02-26     38    96      11766.49    C   39690        50.23       2.825   
2010-02-26      6    94      35521.58    A  202505        47.10       2.561   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-02-05        NaN        NaN        NaN        NaN        NaN  211.096358   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
2010-02-05        NaN        NaN        NaN        NaN        NaN  131.527903   
...               ...        ...        ...        ...        ...         ...   
2010-02-26        NaN        NaN        NaN        NaN        NaN  189.601802   
2010-02-26        NaN        NaN        NaN        NaN        NaN  126.552286   
2010-02-26        NaN        NaN        NaN        NaN        NaN  126.552286   
2010-02-26        NaN        NaN        NaN        NaN        NaN  126.552286   
2010-02-26        NaN        NaN        NaN        NaN        NaN  212.845337   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-02-05         8.106      False              High  
2010-02-05        10.064      False          Moderate  
2010-02-05        10.064      False               Low  
2010-02-05        10.064      False          Moderate  
2010-02-05        10.064      False              High  
...                  ...        ...               ...  
2010-02-26         7.541      False              High  
2010-02-26        13.975      False              High  
2010-02-26         6.548      False          Moderate  
2010-02-26        13.975      False          Moderate  
2010-02-26         7.259      False              High  

[11839 rows x 16 columns]

In [194]:
training_df.loc["2010-08-01":"2010-08-31"]  # Grab values on a range of dates

Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-08-06     28    42       6472.78    A  206302        92.95       3.123   
2010-08-06     34    95      72431.36    A  158114        73.80       2.640   
2010-08-06     40    37       2086.49    A  155083        69.16       2.792   
2010-08-06     38    10        387.23    C   39690        92.95       3.123   
2010-08-06     41    52       1410.66    A  196321        69.21       2.690   
...           ...   ...           ...  ...     ...          ...         ...   
2010-08-27     45    29       4121.43    B  118221        71.36       2.755   
2010-08-27     28    56        459.69    A  206302        93.19       3.129   
2010-08-27     30    79       5909.15    C   42988        86.20       2.619   
2010-08-27     30    97      14782.44    C   42988        86.20       2.619   
2010-08-27     45    34      13559.60    B  118221        71.36       2.755   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  132.614193   
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  190.099003   
...               ...        ...        ...        ...        ...         ...   
2010-08-27        NaN        NaN        NaN        NaN        NaN  182.610406   
2010-08-27        NaN        NaN        NaN        NaN        NaN  126.089290   
2010-08-27        NaN        NaN        NaN        NaN        NaN  211.224176   
2010-08-27        NaN        NaN        NaN        NaN        NaN  211.224176   
2010-08-27        NaN        NaN        NaN        NaN        NaN  182.610406   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-08-06        14.180      False               Low  
2010-08-06         9.816      False              High  
2010-08-06         5.326      False               Low  
2010-08-06        14.180      False               Low  
2010-08-06         7.335      False               Low  
...                  ...        ...               ...  
2010-08-27         8.743      False               Low  
2010-08-27        14.180      False               Low  
2010-08-27         8.099      False               Low  
2010-08-27         8.099      False          Moderate  
2010-08-27         8.743      False          Moderate  

[11603 rows x 16 columns]

---

## Question 3

- Part 1. Slice the Dataframe to only return values from August of 2010

- Part 2. Next, return the number of "Negative" and "Low" sales for the `Sales_Categorical` column from the months of:
    - August, 2010
    - December, 2010
    - January, 2011
    
    


In [205]:
# Part 1 

# Values from Aug 2010
training_df.loc["2010-08-01":"2010-08-31"]

Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-08-06     28    42       6472.78    A  206302        92.95       3.123   
2010-08-06     34    95      72431.36    A  158114        73.80       2.640   
2010-08-06     40    37       2086.49    A  155083        69.16       2.792   
2010-08-06     38    10        387.23    C   39690        92.95       3.123   
2010-08-06     41    52       1410.66    A  196321        69.21       2.690   
...           ...   ...           ...  ...     ...          ...         ...   
2010-08-27     45    29       4121.43    B  118221        71.36       2.755   
2010-08-27     28    56        459.69    A  206302        93.19       3.129   
2010-08-27     30    79       5909.15    C   42988        86.20       2.619   
2010-08-27     30    97      14782.44    C   42988        86.20       2.619   
2010-08-27     45    34      13559.60    B  118221        71.36       2.755   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  132.614193   
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  190.099003   
...               ...        ...        ...        ...        ...         ...   
2010-08-27        NaN        NaN        NaN        NaN        NaN  182.610406   
2010-08-27        NaN        NaN        NaN        NaN        NaN  126.089290   
2010-08-27        NaN        NaN        NaN        NaN        NaN  211.224176   
2010-08-27        NaN        NaN        NaN        NaN        NaN  211.224176   
2010-08-27        NaN        NaN        NaN        NaN        NaN  182.610406   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-08-06        14.180      False               Low  
2010-08-06         9.816      False              High  
2010-08-06         5.326      False               Low  
2010-08-06        14.180      False               Low  
2010-08-06         7.335      False               Low  
...                  ...        ...               ...  
2010-08-27         8.743      False               Low  
2010-08-27        14.180      False               Low  
2010-08-27         8.099      False               Low  
2010-08-27         8.099      False          Moderate  
2010-08-27         8.743      False          Moderate  

[11603 rows x 16 columns]

In [274]:
# Part 2

# Combine values for the Aug 2010, Dec 2010, & Jan 2011
aug_df = training_df.loc["2010-08-01":"2010-08-31"]
dec_df = training_df.loc["2010-12-01":"2010-12-31"]
jan_df = training_df.loc["2011-01-01":"2011-01-31"]

datefiltered_df = pd.concat([aug_df, dec_df, jan_df])

# Check months combined properly
print(pd.concat([aug_df, dec_df, jan_df]).head(3))
print(pd.concat([aug_df, dec_df, jan_df]).tail(3))
print(pd.concat([aug_df, dec_df, jan_df]).count())


            Store  Dept  Weekly_Sales Type    Size  Temperature  Fuel_Price  \
Date                                                                          
2010-08-06     28    42       6472.78    A  206302        92.95       3.123   
2010-08-06     34    95      72431.36    A  158114        73.80       2.640   
2010-08-06     40    37       2086.49    A  155083        69.16       2.792   

            MarkDown1  MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  \
Date                                                                            
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  126.085452   
2010-08-06        NaN        NaN        NaN        NaN        NaN  132.614193   

            Unemployment  isHoliday Sales_Categorical  
Date                                                   
2010-08-06        14.180      False               Low  
2010-08-06         9.816      

In [276]:
# Part 2 continued

salescount = datefiltered_df['Sales_Categorical'].isin(['Negative', 'Low']).sum()
print("Total Negative & Low Sales from Aug'10, Dec'10, & Jan'11:", salescount)

Total Negative & Low Sales from Aug'10, Dec'10, & Jan'11: 21263


---

## Question 4
- Determine the average `Weekly_Sales` for the years of 2010 and 2011 for each store type. 

> Hint, you will need to use groupby and slicing!

In [272]:
# Group A
agroup = training_df.groupby('Type').get_group("A")

avg_sales = agroup['Weekly_Sales'].mean()
reformatted_sales = f"${avg_sales:,.2f}"
print("The Type A store average sales are:", reformatted_sales)


# Group B
bgroup = training_df.groupby('Type').get_group("B")

avg_sales = bgroup['Weekly_Sales'].mean()
reformatted_sales = f"${avg_sales:,.2f}"
print("The Type B store average sales are:", reformatted_sales)


# Group C
cgroup = training_df.groupby('Type').get_group("C")

avg_sales = cgroup['Weekly_Sales'].mean()
reformatted_sales = f"${avg_sales:,.2f}"
print("The Type C store average sales are:", reformatted_sales)


The Type A store average sales are: $20,099.57
The Type B store average sales are: $12,237.08
The Type C store average sales are: $9,519.53


---

# OPTIONAL
> NOTE: Questions 5-8 on are *NOT* required for submission. You can stop here and submit the assignment if needed. 

Great work on these! You've learned to work with Pandas and derive insights from real world data. You've even handled some visualizations and referenced new Python libraries... But there's still more to be done! If you're interested, continue on. We'll deepen our analysis by moving into some basic regressions using a new library, **Statsmodels**.

Remember, Optional and Challenge questions are ***optional.*** Feel free to try them for bonus points or simply come back to them later for additional practice!

> Hint: If you feel like proceeding, we recommend that you check out the documentation for [Statsmodels](https://www.statsmodels.org/stable/index.html) and another graphing library: [Seaborn](https://seaborn.pydata.org/tutorial.html).

---

## Tutorial: Regression Analysis

Welcome to the big leagues! Now that we've learned to slice and bin data with Pandas, let's try some intermediate analysis, shall we? We'll begin with a brief refresher:


### Estimating ("learning") model coefficients

- Coefficients are estimated during the "model fitting" process using the **least squares criterion**.
- Our goal is to find the line (mathematically) which minimizes the **sum of squared residuals** (or "sum of squared errors").

![Estimating coefficients](images/estimating_coefficients.png)

In this diagram:

- The black dots are the **observed values** of x and y.
- The blue line is our **least squares line**.
- The red lines are the **residuals**, which are the *distances between the observed values and the least squares line*.

![Slope-intercept](images/slope_intercept.png)

How do the model coefficients relate to the least squares line?

- $\beta_0$ is the **intercept** (the value of $y$ when $x$=0)
- $\beta_1$ is the **slope** (the change in $y$ divided by change in $x$)


Linear Regression is highly **parametric**, meaning that is relies heavily on the underlying shape of the data. If the data fall into a line, then linear regression will perform well. If the data does **not** fall in line (get it?), then running a linear regression is likely to fail.


Let's try estimating the model coefficients for our Walmart data using the `statsmodels` library.


In [ ]:
### STATSMODELS ###

# Let's import statsmodels and get started!
import statsmodels.formula.api as smf


# Create a fitted model
lm = smf.ols(formula='Weekly_Sales ~ Fuel_Price', data=training_df).fit()

# Print the coefficients
lm.params


### Interpreting model coefficients

How do we interpret the CPI coefficient ($\beta_1$)?

- A "unit" increase in `Fuel_Price` is **associated with** a 5.96 "unit" decrease in Sales.
- This is not a statement of **causation**.

If an increase in `Fuel_Price` was associated with a **increase** in sales, $\beta_1$ would be **positive**.


### Confidence intervals

Statsmodels calculates 95% confidence intervals for our model coefficients, which are interpreted as follows: If the population from which this sample was drawn was **sampled 100 times**, approximately **95 of those confidence intervals** would contain the "true" coefficient.

- We only have a **single sample of data**, and not the **entire population of data**.
- The "true" coefficient is either within this interval or it isn't, but there's no way to actually know.
- We estimate the coefficient with the data we do have, and we show uncertainty about that estimate by giving a range that the coefficient is **probably** within.

> Derived from Quora: [What is a confidence interval in layman's terms?](http://www.quora.com/What-is-a-confidence-interval-in-laymans-terms/answer/Michael-Hochster)

Note: 95% confidence intervals are just a convention. You can also create 90% confidence intervals (which will be more narrow), 99% confidence intervals (which will be wider), or whatever intervals you like.

A closely related concept is **[hypothesis testing](https://www.quora.com/How-do-you-explain-hypothesis-testing-to-a-layman)**.

In [ ]:
### STATSMODELS ###

# Print the confidence intervals for the model coefficients:
lm.conf_int()

---

### Hypothesis testing and p-values

The general process for hypothesis testing is:

- Start with a **null hypothesis** and an **alternative hypothesis** (something that is the opposite of your null).
- Check whether the data supports **rejecting the null hypothesis** or **failing to reject the null hypothesis**.

For model coefficients, here is the "conventional" hypothesis test:

- **null hypothesis:** There is no relationship between TV ads and Sales (and thus $\beta_1$ equals zero)
- **alternative hypothesis:** There is a relationship between TV ads and Sales (and thus $\beta_1$ is not equal to zero)

How do we test this hypothesis?

- The **p-value** is the probability that the relationship we are observing is occurring purely by chance.
- If the 95% confidence interval for a coefficient **does not include zero**, the p-value will be **less than 0.05**, and we will reject the null (and thus believe the alternative).
- If the 95% confidence interval **includes zero**, the p-value will be **greater than 0.05**, and we will fail to reject the null.

In [ ]:
### STATSMODELS ###

#  p-values for the model coefficients. If we just use Fuel Price as our signal, then our p value is very high!
lm.pvalues

---

## Question 5

> Remember, questions 5-8 are *optional*

Run a simple regression analysis using "CPI" as the only signal. According to the analysis, is CPI positively or negatively associated with sales?

In [ ]:
# Now you try!


---

### Tutorial

Want to take our analysis even further? Let's visualize our findings! [Seaborn is a graphing library](https://seaborn.pydata.org/examples/index.html) we can use to create colorful visualizations.

In [ ]:
### SEABORN ###

import seaborn as sns  

# Let's use Seaborn to make a heatmap!
# Hint: See https://www.crazyegg.com/blog/understanding-using-heatmaps-studies for a primer on heatmaps

sns.heatmap(training_df[['Weekly_Sales', 'Temperature', 'CPI', 'Size', 'isHoliday', 'average_markdown']].corr())

---

## Question 6

> Remember, questions 5-8 are *optional*

In the dataset, we have values for CPI, Unemployment, and other entries organized by week. Try graphing "CPI", "Unemployment", and "Fuel Price" as separate line graphs over time.

In [ ]:
# Now you try!


---

## Question 7

> Remember, questions 5-8 are *optional* - attempting them won't hurt your score but you can earn bonus points for trying!

Let's look more granularly by Department. Up until now, we have largely ignored the Store and Department column. 

Now that we have experience grouping by the "Type" column, let's use that knowledge to gain insight about *specific stores and departments.*


- Part 1. There are 99 departments within stores in our dataset. Plot a line graph to show weekly sales by day grouped by department for **departments 1-5** only. What insight can we ascertain about the sales cycle for each department?

- Part 2. Zoom in even further and show the same graph *strictly* for departments 3 and 5. Do they follow the same seasonal pattern as the overall sales trend for the three types of stores?


In [ ]:
# Now you try!


---

# OPTIONAL / CHALLENGE!
> Optional and Challenge Questions are *NOT* required for submission. 

We'll wrap up with an extra *challenge* question to provide a further example of these Python topics in-practice. Here we'll extend our historical analysis in Pandas even further.

**BE WARNED**: Challenge questions are intentionally tough; these questions are designed to integrate advanced concepts and topics not covered in-class. Challenge yourself to continue your learning!



---

## Question 8

> Remember, questions 5-8 are *optional* - attempting them won't hurt your score but you can earn bonus points for trying!

- Part 1. Create a column called `Weekly_Sales_Lag_52` which retuns weekly sales lagged by 52 units (weeks). 

> Hint: You will need to use BOTH groupby and [pandas.shift](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shift.html) to make this column. 

- Part 2. Run a regression analysis using the variables: `Fuel_Price`, `Unemployment`, `CPI`, and `Weekly_Sales_Lag_52`, then print out the p-values for each variable.

In [ ]:
# Now you try!
